src/app/CommandSender.h

```diff
@@ -36,15 +36,69 @@
 #include <system/SystemPacketBuffer.h>
 
 #include <app/Command.h>
+#include <app/MessageDef/CommandPath.h>
+#include <app/MessageDef/StatusElement.h>
 
 #define COMMON_STATUS_SUCCESS 0
 
 namespace chip {
 namespace app {
 
-class CommandSender : public Command, public Messaging::ExchangeDelegate
+class CommandSender final : public Command, public Messaging::ExchangeDelegate
 {
 public:
+    // TODO: Move various CommandResponse related callbacks to this delegate.
+    class Delegate
+    {
+    public:
+        virtual ~Delegate() = default;
+
+        /**
+         * OnResponse will be called when the response from server arrived.
+         *  - When StatusElement is received, aStatus will be filled with the content of the StatusElement, aData will be a
+         * uninitialized TLVReader.
+         *  - When command specific response is received, aStatus will be filled with success status, aData will be initialized with
+         * a struct of command specify response.
+         *
+         * @param[in] apCommandSender: The command sender object that initialized the command transaction, MUST NOT be freed in this
+         * callback.
+         * @param[in] aPath: The command path field in invoke command response.
+         * @param[in] aStatus: The status element, filled with actual StatusElement when a StatusElement is received, when command
+         * specific response is received, aStatus will always be filled with success status code.
+         * @param[in] aData: The command data, or a uninitialized TLVReader if a StatusElement is received.
+         */
+        virtual void OnResponse(const CommandSender * apCommandSender, const CommandPath::Type & aPath,
+                                const StatusElement::Type & aStatus, TLV::TLVReader & aData)
```

Comment:
Given the above comment, `aStatus` should not need to be passed in here.

Instead, convert the TLVReader reference to a pointer, with a null pointer indicating the absence of any received data, and instead, just a success status element received instead.